File: sdks/python/src/fireweave/context.py
"""Canonical evaluation context: bounds, reserved keys, merge, immutability.

Implements ``spec/evaluation-context.schema.json`` v0.1.0:

- Bounds: 128 attributes, 256-byte keys, 4 KiB string values, nesting depth 6,
  64 KiB serialized size.
- Reserved keys: ``targetingKey`` / ``kind`` (configurable) may not appear in
  attributes; ``fireweave.*`` keys are reserved for the SDK except the
  sanctioned carriers (``fireweave.groups``, ``fireweave.groupProperties``
  — rulings 12–13: these are the ONLY permitted ``fireweave.*`` keys).
- Merge order: global -> client -> invocation (later layers win per key).
- Immutability: contexts are frozen; attribute maps are deep-copied on
  construction and exposed via read-only views.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from types import MappingProxyType
from typing import Any, Dict, Iterable, Mapping, Optional, Tuple

from .errors import InvalidContextError, TargetingKeyMissingError
from .types import JsonValue

__all__ = ["ContextLimits", "EvaluationContext", "merge_contexts", "validate_context"]

# Sanctioned fireweave.* carriers (spec/evaluation-context.schema.json,
# orchestrator rulings 12–13): exactly these two — nothing else.
_ALLOWED_FIREWEAVE_KEYS = frozenset(
    {"fireweave.groups", "fireweave.groupProperties"}
)
_DEFAULT_RESERVED_KEYS = ("targetingKey", "kind")


@dataclass(frozen=True)
class ContextLimits:
    """Ratified context bounds (contracts/README.md)."""

    max_attribute_count: int = 128
    max_key_bytes: int = 256
    max_value_bytes: int = 4096
    max_nesting_depth: int = 6
    max_serialized_bytes: int = 65536


def _deep_copy_json(value: Any) -> Any:
    if isinstance(value, dict):
        return {k: _deep_copy_json(v) for k, v in value.items()}
    if isinstance(value, list):
        return [_deep_copy_json(v) for v in value]
    return value


@dataclass(frozen=True)
class EvaluationContext:
    """Immutable Fireweave evaluation context layer.

    ``attributes`` are deep-copied at construction so later mutation of the
    caller's dict cannot leak in, and are exposed through a read-only mapping.
    """

    targeting_key: Optional[str] = None
    attributes: Mapping[str, JsonValue] = field(default_factory=dict)

    def __post_init__(self) -> None:
        copied = _deep_copy_json(dict(self.attributes))
        object.__setattr__(self, "attributes", MappingProxyType(copied))

    def to_dict(self) -> Dict[str, Any]:
        """Plain-dict snapshot (deep copy) of this context."""
        out: Dict[str, Any] = {}
        if self.targeting_key is not None:
            out["targetingKey"] = self.targeting_key
        if self.attributes:
            out["attributes"] = _deep_copy_json(dict(self.attributes))
        return out

    @property
    def vendor_hints(self) -> Dict[str, JsonValue]:
        """``$``-prefixed attributes: vendor pass-through options."""
        return {k: v for k, v in self.attributes.items() if k.startswith("$")}

    @property
    def plain_attributes(self) -> Dict[str, JsonValue]:
        """Attributes minus vendor hints (``$``-prefixed keys)."""
        return {
            k: _deep_copy_json(v)
            for k, v in self.attributes.items()
            if not k.startswith("$")
        }

    @property
    def groups(self) -> Dict[str, str]:
        """Group memberships from ``fireweave.groups`` or plain ``groups``."""
        raw = self.attributes.get("fireweave.groups", self.attributes.get("groups"))
        return dict(raw) if isinstance(raw, dict) else {}

    @property
    def group_properties(self) -> Dict[str, Any]:
        raw = self.attributes.get(
            "fireweave.groupProperties", self.attributes.get("groupProperties")
        )
        return _deep_copy_json(raw) if isinstance(raw, dict) else {}


def merge_contexts(*layers: Optional[EvaluationContext]) -> EvaluationContext:
    """Merge context layers; later layers win per attribute key.

    Order: global -> client -> invocation. ``targeting_key`` from the latest
    layer that sets one wins. Merge is shallow per top-level attribute key.
    """
    targeting_key: Optional[str] = None
    attributes: Dict[str, JsonValue] = {}
    for layer in layers:
        if layer is None:
            continue
        if layer.targeting_key is not None:
            targeting_key = layer.targeting_key
        attributes.update(dict(layer.attributes))
    return EvaluationContext(targeting_key=targeting_key, attributes=attributes)


def _max_depth(value: Any) -> int:
    if isinstance(value, dict):
        return 1 + max((_max_depth(v) for v in value.values()), default=0)
    if isinstance(value, list):
        return 1 + max((_max_depth(v) for v in value), default=0)
    return 0


def _iter_keys(value: Any) -> Iterable[str]:
    if isinstance(value, dict):
        for k, v in value.items():
            yield k
            yield from _iter_keys(v)
    elif isinstance(value, list):
        for v in value:
            yield from _iter_keys(v)


def _iter_string_leaves(value: Any) -> Iterable[str]:
    if isinstance(value, dict):
        for v in value.values():
            yield from _iter_string_leaves(v)
    elif isinstance(value, list):
        for v in value:
            yield from _iter_string_leaves(v)
    elif isinstance(value, str):
        yield value


def validate_context(
    context: EvaluationContext,
    *,
    limits: ContextLimits,
    reserved_keys: Tuple[str, ...] = _DEFAULT_RESERVED_KEYS,
    require_targeting_key: bool = False,
) -> None:
    """Validate a merged context; raise :class:`InvalidContextError` on breach.

    Check order (deterministic, cheap-first): targeting key, reserved keys,
    attribute count, key size, value size, nesting depth, serialized size.
    Validation always happens *before* any backend call.
    """
    if require_targeting_key and not context.targeting_key:
        raise TargetingKeyMissingError()

    attrs = dict(context.attributes)

    reserved = set(reserved_keys)
    for key in attrs:
        if key in reserved:
            raise InvalidContextError("invalid evaluation context")
        if key.startswith("fireweave.") and key not in _ALLOWED_FIREWEAVE_KEYS:
            raise InvalidContextError("invalid evaluation context")

    if len(attrs) > limits.max_attribute_count:
        raise InvalidContextError("context exceeds maximum attribute count")

    for key in _iter_keys(attrs):
        if len(key.encode("utf-8")) > limits.max_key_bytes:
            raise InvalidContextError("context key exceeds maximum size")

    for leaf in _iter_string_leaves(attrs):
        if len(leaf.encode("utf-8")) > limits.max_value_bytes:
            raise InvalidContextError("context value exceeds maximum size")

    if _max_depth(attrs) > limits.max_nesting_depth:
        raise InvalidContextError("context exceeds maximum nesting depth")

    serialized = json.dumps(
        {"targetingKey": context.targeting_key, "attributes": attrs},
        separators=(",", ":"),
        ensure_ascii=False,
        default=str,
    )
    if len(serialized.encode("utf-8")) > limits.max_serialized_bytes:
        raise InvalidContextError("serialized context exceeds maximum size")

File: sdks/go/fireweave/context.go
package fireweave

import (
	"encoding/json"
	"strings"
)

// Reserved attribute keys. "targetingKey" and "kind" are reserved by the
// OpenFeature context model; every key under the "fireweave." prefix is
// reserved for SDK-internal use, with exactly two ratified carve-outs
// (orchestrator rulings 12–14): AttrGroups and AttrGroupProperties are the
// canonical context keys for group targeting and are the ONLY permitted
// "fireweave.*" attributes. Every other "fireweave.*" key is rejected with
// InvalidContext.
const (
	ReservedKeyPrefix = "fireweave."
	// AttrGroups is the canonical context key carrying group memberships
	// (map of group type → group key), e.g.
	// {"organization": "org_123"}. It maps to the vendor "groups" payload
	// field and is never treated as a person property.
	AttrGroups = "fireweave.groups"
	// AttrGroupProperties is the canonical context key carrying per-group
	// properties (map of group type → property map), e.g.
	// {"organization": {"plan": "enterprise"}}. It maps to the vendor
	// "group_properties" payload field and is never treated as a person
	// property.
	AttrGroupProperties = "fireweave.groupProperties"
	// ReservedInvalidContextKey is an internal sentinel attribute injected by
	// the provider's context-guard hook when reserved-key misuse is detected
	// before the Go OpenFeature SDK flattens (and thereby destroys evidence
	// of) the conflict. Its presence always fails validation.
	ReservedInvalidContextKey = "fireweave.invalidContext"

	attrTargetingKey = "targetingKey"
	attrKind         = "kind"
)

// EvaluationContext is Fireweave's structured evaluation context. The
// TargetingKey maps to the vendor distinct_id / cohort id. Attributes may
// nest maps and lists up to the configured depth bound. Contexts are treated
// as immutable: constructors and accessors deep-copy.
type EvaluationContext struct {
	TargetingKey string
	Attributes   map[string]any
}

// NewEvaluationContext builds a context, deep-copying attributes.
func NewEvaluationContext(targetingKey string, attributes map[string]any) EvaluationContext {
	return EvaluationContext{
		TargetingKey: targetingKey,
		Attributes:   deepCopyMap(attributes),
	}
}

// Copy returns a deep copy of the context.
func (c EvaluationContext) Copy() EvaluationContext {
	return EvaluationContext{TargetingKey: c.TargetingKey, Attributes: deepCopyMap(c.Attributes)}
}

// WithGroups returns a copy of the context with the canonical AttrGroups
// ("fireweave.groups") attribute set. This is the idiomatic typed accessor
// ratified by ruling 14: it is sugar over the canonical context key, not a
// separate representation.
func (c EvaluationContext) WithGroups(groups map[string]any) EvaluationContext {
	out := c.Copy()
	if out.Attributes == nil {
		out.Attributes = map[string]any{}
	}
	out.Attributes[AttrGroups] = deepCopyMap(groups)
	return out
}

// WithGroupProperties returns a copy of the context with the canonical
// AttrGroupProperties ("fireweave.groupProperties") attribute set (typed
// sugar per ruling 14; see WithGroups).
func (c EvaluationContext) WithGroupProperties(props map[string]any) EvaluationContext {
	out := c.Copy()
	if out.Attributes == nil {
		out.Attributes = map[string]any{}
	}
	out.Attributes[AttrGroupProperties] = deepCopyMap(props)
	return out
}

// Groups returns a copy of the canonical AttrGroups attribute (nil when
// unset or not a map).
func (c EvaluationContext) Groups() map[string]any {
	g, _ := c.Attributes[AttrGroups].(map[string]any)
	return deepCopyMap(g)
}

// GroupProperties returns a copy of the canonical AttrGroupProperties
// attribute (nil when unset or not a map).
func (c EvaluationContext) GroupProperties() map[string]any {
	g, _ := c.Attributes[AttrGroupProperties].(map[string]any)
	return deepCopyMap(g)
}

func deepCopyMap(m map[string]any) map[string]any {
	if m == nil {
		return nil
	}
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = deepCopyValue(v)
	}
	return out
}

func deepCopyValue(v any) any {
	switch t := v.(type) {
	case map[string]any:
		return deepCopyMap(t)
	case []any:
		out := make([]any, len(t))
		for i, e := range t {
			out[i] = deepCopyValue(e)
		}
		return out
	default:
		return v
	}
}

// MergeContexts merges layers in ascending priority order (earlier layers
// are overridden by later layers, per the OpenFeature merge order
// global → transaction → client → invocation). The targeting key of the
// highest-priority layer that sets one wins. The result is a fresh copy.
func MergeContexts(layers ...EvaluationContext) EvaluationContext {
	out := EvaluationContext{Attributes: map[string]any{}}
	for _, l := range layers {
		if l.TargetingKey != "" {
			out.TargetingKey = l.TargetingKey
		}
		for k, v := range l.Attributes {
			out.Attributes[k] = deepCopyValue(v)
		}
	}
	return out
}

// Limits are the ratified evaluation-context bounds
// (spec/evaluation-context.schema.json).
type Limits struct {
	MaxAttributes      int // maximum top-level attribute count
	MaxKeyBytes        int // maximum attribute key size in bytes
	MaxValueBytes      int // maximum single attribute value size in bytes
	MaxNestingDepth    int // maximum nesting depth of an attribute value
	MaxSerializedBytes int // maximum serialized (JSON) context size in bytes
}

// DefaultLimits returns the canonical bounds: 128 attributes, 256 B keys,
// 4 KiB values, depth 6, 64 KiB serialized.
func DefaultLimits() Limits {
	return Limits{
		MaxAttributes:      128,
		MaxKeyBytes:        256,
		MaxValueBytes:      4096,
		MaxNestingDepth:    6,
		MaxSerializedBytes: 65536,
	}
}

func (l Limits) withDefaults() Limits {
	d := DefaultLimits()
	if l.MaxAttributes <= 0 {
		l.MaxAttributes = d.MaxAttributes
	}
	if l.MaxKeyBytes <= 0 {
		l.MaxKeyBytes = d.MaxKeyBytes
	}
	if l.MaxValueBytes <= 0 {
		l.MaxValueBytes = d.MaxValueBytes
	}
	if l.MaxNestingDepth <= 0 {
		l.MaxNestingDepth = d.MaxNestingDepth
	}
	if l.MaxSerializedBytes <= 0 {
		l.MaxSerializedBytes = d.MaxSerializedBytes
	}
	return l
}

// Canonical validation messages (fixed strings; never interpolate user data).
const (
	msgTargetingKeyMissing = "targeting key missing"
	msgInvalidContext      = "invalid evaluation context"
	msgTooManyAttributes   = "context exceeds maximum attribute count"
	msgKeyTooLarge         = "context key exceeds maximum size"
	msgValueTooLarge       = "context value exceeds maximum size"
	msgTooDeep             = "context exceeds maximum nesting depth"
	msgSerializedTooLarge  = "serialized context exceeds maximum size"
)

// ValidateContext checks a merged context against the bounds and reserved
// key rules. It returns nil when valid, otherwise an InvalidContext error
// whose message never contains attribute values (no PII leakage).
func ValidateContext(c EvaluationContext, limits Limits, requireTargetingKey bool) *Error {
	limits = limits.withDefaults()

	if requireTargetingKey && c.TargetingKey == "" {
		err := NewError(KindInvalidContext, msgTargetingKeyMissing, nil)
		err.TargetingKeyMissing = true
		return err
	}
	for k, v := range c.Attributes {
		if k == attrTargetingKey || k == attrKind {
			return NewError(KindInvalidContext, msgInvalidContext, nil)
		}
		if strings.HasPrefix(k, ReservedKeyPrefix) {
			// Ratified carve-out (rulings 12–14): fireweave.groups and
			// fireweave.groupProperties are the only permitted fireweave.*
			// keys, and both must carry map values.
			if k != AttrGroups && k != AttrGroupProperties {
				return NewError(KindInvalidContext, msgInvalidContext, nil)
			}
			if _, ok := v.(map[string]any); !ok {
				return NewError(KindInvalidContext, msgInvalidContext, nil)
			}
		}
	}
	if len(c.Attributes) > limits.MaxAttributes {
		return NewError(KindInvalidContext, msgTooManyAttributes, nil)
	}
	for k, v := range c.Attributes {
		if len(k) > limits.MaxKeyBytes {
			return NewError(KindInvalidContext, msgKeyTooLarge, nil)
		}
		if valueSize(v) > limits.MaxValueBytes {
			return NewError(KindInvalidContext, msgValueTooLarge, nil)
		}
		if depthOf(v) > limits.MaxNestingDepth {
			return NewError(KindInvalidContext, msgTooDeep, nil)
		}
	}
	if serializedSize(c) > limits.MaxSerializedBytes {
		return NewError(KindInvalidContext, msgSerializedTooLarge, nil)
	}
	return nil
}

func valueSize(v any) int {
	if s, ok := v.(string); ok {
		return len(s)
	}
	b, err := json.Marshal(v)
	if err != nil {
		return 0
	}
	return len(b)
}

// depthOf reports the nesting depth of a value: scalars are depth 1, each
// enclosing map or list adds one level.
func depthOf(v any) int {
	switch t := v.(type) {
	case map[string]any:
		max := 0
		for _, e := range t {
			if d := depthOf(e); d > max {
				max = d
			}
		}
		return 1 + max
	case []any:
		max := 0
		for _, e := range t {
			if d := depthOf(e); d > max {
				max = d
			}
		}
		return 1 + max
	default:
		return 1
	}
}

func serializedSize(c EvaluationContext) int {
	b, err := json.Marshal(map[string]any{
		"targetingKey": c.TargetingKey,
		"attributes":   c.Attributes,
	})
	if err != nil {
		return 0
	}
	return len(b)
}
